services/motion/builtin/moveonglobe.go

```diff
@@ -0,0 +1,314 @@
+	componentName string,
+) (motionplan.Plan, error) {
+	inputs, err := kb.CurrentInputs(ctx)
+	if err != nil {
+		return make(motionplan.Plan, 0), err
+	}
+	// TODO: this is really hacky and we should figure out a better place to store this information
+	if len(kb.Kinematics().DoF()) == 2 {
+		inputs = inputs[:2]
+	}
+	planRequest.StartConfiguration = map[string][]referenceframe.Input{componentName: inputs}
+
+	return motionplan.PlanMotion(ctx, planRequest)
+}
+
+func startPollingForReplan(
+	ctx context.Context,
+	period time.Duration,
+	responseChan chan replanResponse,
+	backgroundWorkers *sync.WaitGroup,
+	fn func(context.Context,
+	) replanResponse,
+) {
+	backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ticker := time.NewTicker(period)
+		defer ticker.Stop()
+		for {
+			// this ensures that if the context is cancelled we always
+			// return early at the top of the loop
+			if err := ctx.Err(); err != nil {
+				return
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				response := fn(ctx)
+				if response.err != nil || response.replan {
+					responseChan <- response
+					return
+				}
+			}
+		}
+	}, backgroundWorkers.Done)
+}
+
+func (ps *planSession) start(
+	motionCfg *motion.MotionConfiguration,
+	kb kinematicbase.KinematicBase,
+	ms *builtIn,
+	movementSensor movementsensor.MovementSensor,
+	destination *geo.Point,
+	replanCount int,
+) {
+	positionPollingPeriod := time.Duration(1000/motionCfg.PositionPollingFreqHz) * time.Millisecond
+	obstaclePollingPeriod := time.Duration(1000/motionCfg.ObstaclePollingFreqHz) * time.Millisecond
+
+	// helper function to manage polling functions
+	// spawn two goroutines that each have the ability to trigger a replan
+	if positionPollingPeriod > 0 {
+		startPollingForReplan(
+			ps.cancelCtx,
+			positionPollingPeriod,
+			ps.positionChan,
+			ps.backgroundWorkers,
+			func(ctx context.Context) replanResponse {
+				if replanCount == 2 {
+					return replanResponse{replan: true}
+				}
+				return replanResponse{}
+			})
+	}
+	if obstaclePollingPeriod > 0 {
+		startPollingForReplan(
+			ps.cancelCtx,
+			obstaclePollingPeriod,
+			ps.obsticleChan,
+			ps.backgroundWorkers,
+			func(ctx context.Context) replanResponse {
+				if replanCount == 1 {
+					return replanResponse{replan: true}
+				}
+				return replanResponse{}
+			})
+	}
+
+	// spawn function to execute the plan on the robot
+	ps.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		if err := ms.execute(ps.cancelCtx, kb, ps.plan); err != nil {
+			ps.executionChan <- executeResponse{err: err}
+			return
+		}
+
+		// the plan has been fully executed so check to see if the GeoPoint we are at is close enough to the goal.
+		success, err := arrivedAtGoal(
+			ps.cancelCtx,
+			movementSensor,
+			destination,
+			motionCfg.PlanDeviationMM,
+		)
+		if err != nil {
+			ps.executionChan <- executeResponse{err: err}
+			return
+		}
+
+		ps.executionChan <- executeResponse{success: success}
```

Comment:
This solves this deadlock problem in the current PR: https://github.com/viamrobotics/rdk/pull/2791/files#r1311918579